samza-core/src/main/java/org/apache/samza/storage/StorageRecovery.java

```diff
@@ -61,13 +62,12 @@
   private Config jobConfig;
   private int maxPartitionNumber = 0;
   private File storeBaseDir = null;
-  private HashMap<String, SystemStream> changeLogSystemStreams = new HashMap<String, SystemStream>();
-  private HashMap<String, StorageEngineFactory<?, ?>> storageEngineFactories = new HashMap<String, StorageEngineFactory<?, ?>>();
-  private HashMap<String, SystemFactory> systemFactories = new HashMap<String, SystemFactory>();
-  private HashMap<String, SystemAdmin> systemAdmins = new HashMap<String, SystemAdmin>();
-  private Map<String, ContainerModel> containers = new HashMap<String, ContainerModel>();
-  private List<TaskStorageManager> taskStorageManagers = new ArrayList<TaskStorageManager>();
+  private HashMap<String, SystemStream> changeLogSystemStreams = new HashMap<>();
+  private HashMap<String, StorageEngineFactory<?, ?>> storageEngineFactories = new HashMap<>();
+  private Map<String, ContainerModel> containers = new HashMap<>();
+  private List<TaskStorageManager> taskStorageManagers = new ArrayList<>();
   private Logger log = LoggerFactory.getLogger(StorageRecovery.class);
+  private SystemAdmins systemAdmins = null;
```

Comment:
If this is started and stopped in run, do we need to save it as a field?

Reply:
The reason it is kept as a field is that this SystemAdmin is instantiated in `getTaskStorageManagers()` and it is shared by multiple `taskStorageManager`. I don't have a good way to avoid saving it as a field. What do you think?

Reply:
Fixed as suggested.